Add ToS Acceptance Tracking to login, signup and checkout

claudiucelfilip opened the pull request:
#### Changes proposed in this Pull Request

More info: pau2Xa-3my-p2

This adds sends the ToS Acceptance payload in the follow scenarios:
- log-in
- sign-up
- checkout

The payload is included under the `tos` key and contains:
- `path` - the route where the ToS was accepted
- `locale` - selected locale code
- `viewport` - width and height of the viewport

#### Testing instructions

* Sandbox `public-api.wordpress.com`
* Apply the patches from D72596-code and D72595-code so that the ToS payload gets processed
* Open the Network tab with Preserve Logs enabled
* Go through the following scenarios:
  * Log In
    * go to `/log-in`
    * log in with username and password
    * make sure the request to `wp-login.php?action=login-endpoint` contains the `tos` property
  * Log In email link
    * go to `/log-in/link`
    * enter email
    * make sure the request to `/auth/send-login-emai` contains the `tos` property   
  * Social Log In
    * for this, you'll need to configure your calypso to run behind `wordpress.com`
    * go to `/log-in`
    * log in with Google or Apple
    * make sure the request to `wp-login.php?action=social-login-endpoint` contains the `tos` property
  * Sign Up
    * go to `/start`
    * enter email, username, and password
    * make sure the request to `/users/new` contains the `tos` property
  * Social Up
    * go to `/start`
    * register with a new Google or Apple account
    * make sure the request to `/users/social/new` contains the `tos` property
  * Social Sign Up with existing account
    * for this, you'll need to configure your calypso to run behind `wordpress.com`
    * go to `/start`
    * register with an already used Google or Apple account
    * make sure the request to `/users/social/new` contains the `tos` property
  * Checkout
    * proceed to the checkout page 
    * fill in the details
    * click on Pay
    * make sure the request to `/me/transactions` contains the `tos` property
  * Checkout PayPal
    * proceed to the checkout page 
    * fill in the details
    * choose PayPal as Payment Method and submit
    * make sure the request to `/me/paypal-express-url` contains the `tos` property

claudiucelfilip committed:
Add small library to generate the ToS acceptance payload

claudiucelfilip committed:
Adds ToS acceptance payload to the login, register and checkout requests

claudiucelfilip committed:
Harden the ToS payload library to work on non-window environments

claudiucelfilip committed:
Update checkout related tests to expect the ToS tracking properties

sirbrillig commented on client/lib/tos-acceptance-tracking/index.js:
```diff
@@ -0,0 +1,19 @@
+import { getLocaleSlug } from 'calypso/lib/i18n-utils';
+
+export default function () {
```
Minor, but it may help future developers to find this function if it has a name instead of being anonymous. Since it's being imported everywhere as `getToSAcceptancePayload` that would probably be a good name to give it here too.

claudiucelfilip replied:
The function is now named. Thanks for pointing this out!

sirbrillig commented on client/my-sites/checkout/composite-checkout/lib/submit-wpcom-transaction.ts:
```diff
@@ -57,5 +59,14 @@ export default async function submitWpcomTransaction(
+		mapRecordKeysRecursively(
+			{
+				...payload,
+				tos: getToSAcceptancePayload(),
```
This works fine, but I'm wary of adding properties to transaction data in this function because the data is not covered by types if it's changed once it gets here. Once we start modifying it, people are more likely to put additional changes to the data here.

Is there any possibility that we could add `tos` to the data just before this function is called? We even have functions to centralize that logic: `prepareRedirectTransaction()`, `prepareFreePurchaseTransaction()`, etc. or probably even better: `createTransactionEndpointRequestPayload()`.

claudiucelfilip replied:
I think you make a fair point there and using `createTransactionEndpointRequestPayload()` looks like the best place to add the `tos` data. Thanks!

sirbrillig replied:
👏

sirbrillig replied:
Thanks! This inspired me to add comments to this function to make it more clear for future developers: https://github.com/Automattic/wp-calypso/pull/59888

sirbrillig reviewed (commented):
Nice job adding tests! Thank you! I have one particular concern which I've put inline. Let me know what you think.

claudiucelfilip committed:
Name ToS acceptance payload function for better searchability

claudiucelfilip committed:
Use transaction payload modifier function to add the ToS acceptance data to the request

sirbrillig reviewed (approved):
I haven't tested any of the signup or other changes, but from a checkout perspective, this looks good and works as described:
<img width="199" alt="Screen Shot 2022-01-07 at 1 47 56 PM" src="https://user-images.githubusercontent.com/2036909/148592104-0cd53f2c-bba8-4f52-84b0-c1b382f31c6c.png">

sirbrillig commented:
Oh, one other thing. You may want to add this behavior to the PayPal processor as well, which uses a different transactions endpoint: https://github.com/Automattic/wp-calypso/blob/bde1ea4840bdd8136b3f4938225f7a00820369cf/client/my-sites/checkout/composite-checkout/lib/paypal-express-processor.ts#L123

claudiucelfilip committed:
Add ToS payload to PayPal express checkout

claudiucelfilip committed:
Update PayPal test expecations to include the ToS payload

claudiucelfilip commented:
> Oh, one other thing. You may want to add this behavior to the PayPal processor as well, which uses a different transactions endpoint:
> 
> https://github.com/Automattic/wp-calypso/blob/bde1ea4840bdd8136b3f4938225f7a00820369cf/client/my-sites/checkout/composite-checkout/lib/paypal-express-processor.ts#L123

@sirbrillig Thanks for reviewing and catching this! 
I added the ToS payload to the PayPal processor request as well and updated the test instructions.

sirbrillig reviewed (approved):
Thanks for adding that! Yeah, I nearly forgot about the PayPal code too. I'm adding comments and a little clean-up in https://github.com/Automattic/wp-calypso/pull/59888 to help us all remember in the future.

claudiucelfilip committed:
Update failing tos payload for unit tests

claudiucelfilip committed:
Add the ToS package to the Magic Link login request

niranjan-uma-shankar reviewed (approved):
Tested along with D72595-code and works well :shipit:
